components/rollup_node/block_gen.md

```diff
@@ -0,0 +1,147 @@
+
+Sequencer blocks are submitted by the sequencer and target _future_ epochs which satisfy the following two conditions:
+
+1. Target epoch is larger than the current L1 block number.
+2. Target epoch is less than the current L1 block number PLUS `sequencing_window_size`.
+
+The ability for the sequencer to append sequencer blocks to future epochs allows the sequencer to predict and influence the state of the epoch before the L1 chain has mined it. This property is what enables _fast transaction confirmations_ via the sequencer replicator.
+
+### Epoch Structure
+
+Each epoch contains **1** deposit block and zero to many sequencer blocks. For epoch `n` the deposit block is generated using L1 block number `n - (sequencing_window_size + 1)`. The epoch's sequencer blocks are contained in any of the L1 blocks ranging from `n - sequencing_window_size` to `n`. This a range of blocks is called the "sequencing window".
+
+The following diagram demonstrates the correspondence between L1 blocks and rollup blocks (ie L2 blocks):
+
+![Sequencer block generation diagram](../../assets/sequencer-block-gen.svg)
+
+## Deposit Blocks in Depth
+
+For every L1 block (after the rollup's genesis) an L2 deposit block is created. These deposit blocks contain both a `ContextDeposit` and any number of `UserDeposit`s. Context deposits set contextual information about the latest L1 block (eg. `blockHash` and `timestamp`), and user deposits are L1 user initiated L2 transactions which guarantee liveness of the rollup chain even with a censoring sequencer.
+
+The types of deposit blocks are as follows:
+
+```python
+class DepositBlock:
+    deposits: List[Deposit]
+
+class Deposit:
+    feedIndex: uint64
+    GasLimit:  uint64
+
+class UserDeposit(Deposit):
+    isEOA:       bool
+    l1TxOrigin:  Address
+    target:      Address
+    data:        bytes
+
+class ContextDeposit(Deposit):
+    blockHash:   bytes32
+    blockNumber: uint64
+    timestamp:   uint64
+    baseFee:     uint64
+```
+
+### Deposit Block Generation
+
+The function `generate_deposit_block(l1_block_number)` is defined as:
+
+- Generate the `ContextDeposit` using the L1 block body. Specifically the L1 block's `blockHash`, `blockNumber`, `timestamp`, and `baseFee`.
+- Get all events emitted by the `DepositFeed` contract at block `l1_block_number`.
+    - For each event, generate a new `UserDeposit` based on the emitted calldata. All fields should be emitted by DepositFeed contract.
+
+#### **Footnote**: The DepositFeed ensures that no matter the size of the event it is possible to prove the deposit data to the fraud proof. See the DepositFeed spec for details.
+
+## Sequencer Blocks in Depth
+
+Sequencer blocks contain a majority of the transactions submitted to the rollup chain. Sequencer blocks can also influence the current timestamp of L2 which is exposed to contracts & is used in the EIP1559 fee pricing calculations.
+
+Sequencer blocks are of the following form:
+
+```python
+class SequencerBlock:
```

Comment:
Currently this type does not allow for/delineate multiple L2 blocks. Presumably we could just add an array of `num_transactions, time_increase` which would say how many of the `transactions` to pull into the next L2 block, and how much to add to the previous L2 block's timestamp.

Reply:
Ah so I intentionally left out the batch submitter encoding scheme because it was going to bloat the scope a bit. My intention is that after you decode the batches that it results in an array of `SequencerBlock`s (aka `SequencerFeedElements`) and those eventually become blocks. Then we can work on optimizing the encoding scheme in another PR. That said, I do think that there is a very simple encoding scheme which should work & I'll write up my strawman soon

Reply:
Ah so we might actually have too many terms at this point which makes it confusing.

A "batch" is just a single transaction from the sequencer to the sequencer feed with any number of blocks that it can fit. What makes this "batch" term a bit more confusing now is that we can actually submit transactions to the sequencer feed that include partial blocks. So you don't need to submit a batch that contains a specific number of blocks but instead submit a batch with as much info as possible and then start the next batch off continuing where you left off.

The fact that a "batch" is just the sequencer appending to the sequencer feed makes me a little less into the term. Hmm... one of the major edits I think we'll have to make once we've sketched out the design is do a mega bikeshed to determine what terms are useful and what terms are overkill.